source/common/quic/envoy_quic_connection_debug_visitor_factory_interface.h

```diff
@@ -15,23 +14,29 @@
 namespace Envoy {
 namespace Quic {
 
-class EnvoyQuicConnectionDebugVisitorFactoryInterface : public Config::TypedFactory {
+class EnvoyQuicConnectionDebugVisitorFactoryInterface {
 public:
-  std::string category() const override { return "envoy.quic.connection_debug_visitor"; }
-
-  void setContext(Envoy::ProcessContextOptRef context) { context_ = context; }
+  virtual ~EnvoyQuicConnectionDebugVisitorFactoryInterface() = default;
 
   // Returns a debug visitor to be attached to a Quic Connection.
   virtual std::unique_ptr<quic::QuicConnectionDebugVisitor>
-  createQuicConnectionDebugVisitor(quic::QuicSession* session,
+  createQuicConnectionDebugVisitor(Event::Dispatcher& dispatcher, quic::QuicSession* session,
                                    const StreamInfo::StreamInfo& stream_info) PURE;
-
-protected:
-  Envoy::ProcessContextOptRef context_;
```

Comment:
Your extension's factory can get this in `createFactory()` with `listener_context.getTransportSocketFactoryContext().serverFactoryContext().processContext()`.

Reply:
Why do we need to pass down a ListenerFactoryContext with this change? QuicDebugVisitor can be used on both server and client. Can we make the interface depend on perspective as less as possible?

Reply:
I needed access to the correct stats scope (listener scope in this case) and I knew you needed the processContext.

I hadn't thought about the upstream side; that isn't currently implemented. We could make a new context type that returns the appropriate scope depending on context if/when upstream support for debug visitors is added.

Reply:
That's a widely used stats scope in listeners, can you access it via `ListenerConfig::listenerScope()` instead? Anyway I would prefer adding individual parameter to the debug visitor factory interfaces upon need, even though this will be an API breaking change. (Maybe wrapping all parameters into a struct to avoid future API breaking changes, but that's not mandatory).

Reply:
> Maybe wrapping all parameters into a struct to avoid future API breaking changes, but that's not mandatory

That's what all of the factory contexts are in envoy; a way to group the parameters to the object factories. It's an interface instead of a struct, but essentially the same thing. It's the way this is handled for most/all envoy extensions.

Reply:
In your case, I suspect you don't event need to make your own context interface/struct, you can just access listener_config.listenerScope() during debug visitor creation.

Reply:
I think what I have now is the best way to handle this until someone adds upstream debug-visitor support. The idea behind the various factory contexts is to pass everything an extension may need, so that we don't have to modify the code that creates it whenever a new extension is added that needs another common interface. We can re-visit this if/when upstream support is added.

Reply:
cc @wlampert for API breaking change